File: 백준/Silver/1991. 트리 순회/트리 순회.py
# 다시보기
import sys
input = sys.stdin.readline
n = int(input())
# tree를 딕셔너리 형태로 선언 => 트리의 노드 값이 A, B, C, D이므로
tree = {}

for _ in range(n):
    root, left, right = input().split()
    # 딕셔너리 tree 루트값에 왼쪽자식노드, 오른쪽자식노드 값들을 차례로 넣음
    tree[root] = [left, right]

# 전위 순회 함수, 루트-왼쪽-오른쪽
def preOrder(node):
    if node == '.':             # 자식 노드가 없는 경우
        return
    print(node, end = '')       # 현재 노드 출력
    preOrder(tree[node][0])     # 왼쪽 탐색
    preOrder(tree[node][1])     # 오른쪽 탐색

# 중위 순회 함수, 왼쪽-루트-오른쪽
def inOrder(node):
    if node == '.':
        return
    inOrder(tree[node][0])      # 왼쪽 탐색
    print(node, end = '')       # 현재 노드 출력
    inOrder(tree[node][1])      # 오른쪽 탐색

# 후위 순회 함수, 왼쪽-오른쪽-루트
def postOrder(node):
    if node == '.':
        return
    postOrder(tree[node][0])    # 왼쪽 탐색
    postOrder(tree[node][1])    # 오른쪽 탐색
    print(node, end='')         # 현재 노드 출력

# 루트 노드가 A이므로 전위, 중위, 후위 순회할 때마다 맨 처음 루트 노드인 A를 넣어준다
preOrder('A')
print()
inOrder('A')
print()
postOrder('A')

File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

// char 알파벳과 왼쪽 노드, 오른쪽 노드를 가진 노드 객체를 만들어 구현
public class Main {
	static class Node{
		char data;
		Node left, right;
		
		public Node(char data, Node left, Node right) {
			super();
			this.data = data;
			this.left = left;
			this.right = right;
		}
	}
	
	static int N;
	// 루트노드를 저장
	static Node head = new Node('A', null, null);

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		N = Integer.parseInt(br.readLine());
		
		for(int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			char root = st.nextToken().charAt(0);
			char left = st.nextToken().charAt(0);
			char right = st.nextToken().charAt(0);
			insertNode(head, root, left, right);
		}
		
		preOrder(head);
		System.out.println();
		inOrder(head);
		System.out.println();
		postOrder(head);
	}
	
	public static void insertNode(Node tmp, char root, char left, char right) {
		// 루트노드인지 확인
		// 만약 루트노드이고 left나 right 값이 '.'이 아니라면 해당 부분에 새로운 노드 생성
		if(tmp.data == root) {
			tmp.left = (left == '.' ? null : new Node(left, null, null));
			tmp.right = (right == '.' ? null : new Node(right, null, null));
		}
		else {
			if(tmp.left != null) insertNode(tmp.left, root, left, right);
			if(tmp.right != null) insertNode(tmp.right, root, left, right);
		}
	}
	
	public static void preOrder(Node node) {
		if(node == null) return;
		System.out.print(node.data);
		preOrder(node.left);
		preOrder(node.right);
	}
	
	public static void inOrder(Node node) {
		if(node == null) return;
		inOrder(node.left);
		System.out.print(node.data);
		inOrder(node.right);
	}
	
	public static void postOrder(Node node) {
		if(node == null) return;
		postOrder(node.left);
		postOrder(node.right);
		System.out.print(node.data);
	}
}